### PR TITLE
feat: fedimint metrics

### PR DESCRIPTION
**Pull request by douglaz**
Export LN metrics from the server in [prometheus](https://prometheus.io/) format. More metrics can be added in future.

This supports the monitoring and administration of servers.

It will look at a local database. Parameters are similar to `dbtool`

### CODE REVIEW

**Review by EthnTuttle — COMMENTED** [review]
Does prometheus format mean we have to run prometheus beside it? Or will this pipe directly into Grafana?

**Comment by douglaz** [comment]
> Does prometheus format mean we have to run prometheus beside it? Or will this pipe directly into Grafana?

You can use [Grafana Agent ](https://grafana.com/docs/agent/latest/) as a lightweight alternative do load the metrics in graphana

**Comment by EthnTuttle** [comment]
Does this PR enable us to test locally with a standard Grafana install? If so, could you provide a short HOW TO? This is my first time working with the combo so would appreciate any guidance on how this work. Happy with RTFM links as well.

**Comment by douglaz** [comment]
The easiest way is to create an account on https://grafana.com/, then `Add new connection`, pick `Hosted Prometheus metrics` and `Via Grafana Agent`. Follow the instructions to create a new config and change
```yaml
- targets: ['localhost:9100']
```
to
```yaml
- targets: ['localhost:3000']
```

Then you can build dashboards using the `grafanacloud-xxx-prom`  `data source`.

But of course before that start the metrics with something like:
```bash
for i in 0; do fedimint-metrics monitor --database $FM_DATA_DIR/server-$i/database --cfg-dir $FM_DATA_DIR/server-$i --password pass$i;done
```

And to make some lightning payments, you can run
```bash
fedimint-load-test-tool load-test --generate-invoice-with cln-lightning-cli
```

**Comment by m1sterc001guy** [comment]
Cool! I like this approach and was along the same lines as what I was thinking. I will look at the code closer tomorrow but I had some higher level questions:

1. Could you add the above explanation to a README for the crate? In my opinion, your description on how to load it into Grafana is essential knowledge for running this code. 
2. Does Grafana run locally when running it in the browser? I'm curious how it is able to access `localhost` when using grafana.com
3.  Is it possible to make this API authenticated? Some guardians might want to run a web-accessible dashboard but not want the whole world able to query it

File: fedimint-metrics/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+    process_update_time_ms: U64IntGauge,
+}
+
+impl PrometheusMetrics {
+    pub fn new() -> anyhow::Result<Self> {
+        let registry = Registry::new();
+
+        let ln_incoming_offer =
+            U64IntGauge::new("ln_incoming_offer", "contracts::IncomingContractOffer")?;
+        registry.register(Box::new(ln_incoming_offer.clone()))?;
```
Comment (EthnTuttle) [inline fedimint-metrics/src/main.rs:315]: Why does this require `Box::new`?

Comment (douglaz) [reply fedimint-metrics/src/main.rs:315]: Sometimes you have many implementations of a trait and using generics can be tricky, then you alloc on the heap with box and things are easier to implement/use.

**Comment by douglaz** [comment]
> Cool! I like this approach and was along the same lines as what I was thinking. I will look at the code closer tomorrow but I had some higher level questions:
> 
> 1. Could you add the above explanation to a README for the crate? In my opinion, your description on how to load it into Grafana is essential knowledge for running this code.
Yes

> 2. Does Grafana run locally when running it in the browser? I'm curious how it is able to access `localhost` when using grafana.com
Grafana is running remotely. What is running locallly is `Grafana Agent` which pushes the metrics to `grafana.com` endpoint

> 3. Is it possible to make this API authenticated? Some guardians might want to run a web-accessible dashboard but not want the whole world able to query it
This won't happen. `Grafana Agent` will run locally and push the metrics to some remote endpoint, so you don't need to open this api port to the world, it can always listen on `localhost`

File: fedimint-metrics/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+
+async fn run_update_metrics_loop(
+    state: Arc<RwLock<State>>,
+    args: MonitorArgs,
+) -> anyhow::Result<()> {
+    let cfg = read_server_config(&args.password, args.cfg_dir)?;
+
+    loop {
+        let start = fedimint_core::time::now();
+        let read_only = RocksDbReadOnly::open_read_only(args.database.clone())?;
```
[the remark this answers is not in the document]
Comment (douglaz) [reply fedimint-metrics/src/main.rs:92]: Yes, being able to update incrementally the metrics is a big feature if it runs inside the process. I'll try to prototype it to see how it goes.

[the remark this answers is not in the document]
Comment (douglaz) [reply fedimint-metrics/src/main.rs:92]: Created https://github.com/fedimint/fedimint/pull/2681 but I'm finding it very tricky to count things just once, I couldn't figure out the lifecycle of some events

**Comment by douglaz** [comment]
@elsirion When you have time take a look at this and https://github.com/fedimint/fedimint/pull/2681

**Comment by douglaz** [comment]
Closed in favor of https://github.com/fedimint/fedimint/pull/2681

**Pull request closed without merging**